jobs/kpi-forecasting/kpi_forecasting/tests/test_funnel_forecast.py

```diff
@@ -0,0 +1,1603 @@
+        return {"yhat": {0: output}}
+
+
+def mock_build_model(segment_settings, parameters):
+    """mocks the FunnelForecast build_model method"""
+    return MockModel(
+        **parameters,
+    )
+
+
+def mock_get_crossvalidation_metric(m, *args, **kwargs):
+    """mocks the FunnelForecast get_crossvalidation_metric
+    method, meant to be used with MockModel"""
+    return m.value  # value atrribute in MockModel
+
+
+def mock_aggregate_forecast_observed(
+    forecast_df, observed_df, period, numpy_aggregations, percentiles
+):
+    """Mocks the aggregate_forecast_observed function defined in ProphetForecast
+    and inherited in FunnelForecast.
+    This function is tested extensively in test_prophet_forecast
+    so we can make dummy outputs for tests related to it"""
+
+    # add dummy columns where aggregated metrics woudl go
+    percentile_columns = [f"p{el}" for el in percentiles]
+    output_forecast_df = forecast_df.copy()
+    output_forecast_df[numpy_aggregations + percentile_columns] = 0
+    return output_forecast_df, observed_df.copy()
+
+
+def test_combine_forecast_observed(mocker, forecast):
+    """tests the _combine_forecast_observed method"""
+    mocker.patch.object(
+        forecast, "_aggregate_forecast_observed", mock_aggregate_forecast_observed
+    )
+
+    forecast_df = pd.DataFrame(
+        {
+            "submission_date": [
+                pd.to_datetime("2124-01-01").date(),
+                pd.to_datetime("2124-01-02").date(),
+            ],
+        }
+    )
+
+    observed_df = pd.DataFrame(
+        {
+            "submission_date": [
+                pd.to_datetime("2124-01-01").date(),
+                pd.to_datetime("2124-01-02").date(),
+            ],
+            "a": ["A1", "A1"],
+            "value": [5, 6],
+        }
+    )
+
+    numpy_aggregations = ["mean"]
+    percentiles = [10, 50, 90]
+
+    output_df = forecast._combine_forecast_observed(
+        forecast_df=forecast_df,
+        observed_df=observed_df,
+        period="period",
+        numpy_aggregations=numpy_aggregations,
+        percentiles=percentiles,
+        segment={"a": "A1"},
+    )
+
+    # mean was renamed to value, percentiles to high, medium, low
+    forecast_df[["value", "value_low", "value_mid", "value_high"]] = 0
+    forecast_df["a"] = "A1"  # this column is already present in observed
+
+    forecast_df["source"] = "forecast"
+    observed_df["source"] = "historical"
+
+    # concat in same order to make our lives easier
+    expected = pd.concat([observed_df, forecast_df])
+    assert set(expected.columns) == set(output_df.columns)
+    pd.testing.assert_frame_equal(output_df, expected[output_df.columns])
+
+    # should not be any nulls outside the metric column
+    non_metric_columns = [
+        el
+        for el in output_df.columns
+        if el not in ["value", "value_low", "value_mid", "value_high"]
+    ]
+    assert not pd.isna(output_df[non_metric_columns]).any(axis=None)
```

Comment:
Why filter to `non_metric_columns` here? I don't think there should be null values in any of the columns.

Reply:
Currently for the observed data the percentile columns ("value_low", "value_mid", "value_high") are NaN.  Is that expected?

Reply:
Ahh, you right, whoops